### PR TITLE
Immmutable state based skyjo implementation

### PR DESCRIPTION
**Pull request by urolyi**
The current state of my immutable state based skyjo implementation.

### CODE REVIEW

**Comment by urolyi1** [comment]
Code could probably be cleaned up a bit, but there are tests and I ran some sims to check that it seems to be working with random players. I'm probably going to start actually writing up the training/data generation loops now as well as addressing any comments/feedback.

File: skyjo_immutable.py [hunk 1]
```diff
@@ -0,0 +1,513 @@
...
+    row_idx: int | None = dataclasses.field(default=None)
+    col_idx: int | None = dataclasses.field(default=None)
+
+
+@dataclasses.dataclass(slots=True, frozen=True)
+class Hand:
+    is_cleared: np.array = dataclasses.field(
+        default_factory=lambda: np.zeros(NUM_ROWS * NUM_COLUMNS)
+    )
+    is_flipped: np.array = dataclasses.field(
```
Comment (noahbkim) [inline skyjo_immutable.py:52]: is it possible to just make this an array of `bool`? same for above? could you combine them into a single set of flags i.e. `state & 0b01` is cleared and `state & 0b10` is flipped?

File: skyjo_immutable.py [hunk 2]
```diff
@@ -0,0 +1,513 @@
...
+    def visible_points(self) -> int:
+        return self.cards[
+            np.where((self.is_cleared == 0) & (self.is_flipped == 1))
+        ].sum()
+
+    def has_all_revealed(self) -> bool:
+        return self.is_flipped.sum() == NUM_COLUMNS * NUM_ROWS
+
+    def total_points(self) -> int:
+        return self.cards[np.where((self.is_cleared == 0))].sum()
```
Comment (noahbkim) [inline skyjo_immutable.py:144]: redundant `()`

File: skyjo_immutable.py [hunk 3]
```diff
@@ -0,0 +1,513 @@
...
+        next_discarded_card_counts[self.top_card + 2] -= 1
+        next_discarded_card_counts[card + 2] += 1
+        return dataclasses.replace(
+            self, discarded_card_counts=next_discarded_card_counts, top_card=card
+        )
+
+
+@dataclasses.dataclass(slots=True, frozen=True)
+class ImmutableSkyjoState:
+    num_players: int = dataclasses.field()
```
Comment (noahbkim) [inline skyjo_immutable.py:263]: unneeded `dataclasses.field`? same below

File: skyjo_immutable.py [hunk 4]
```diff
@@ -0,0 +1,513 @@
...
+    def setup_round(self) -> Self:
+        next_deck = Deck(_cards=DECK, _rng=self.deck._rng).shuffle()
+        next_hands, next_deck = next_deck.deal_hands(self.num_players)
+        top_card, next_deck = next_deck.draw_top_card()
+        next_hands = [hand.initial_flips() for hand in next_hands]
+        starting_player = self.round_ending_player
+        if self.round_ending_player is None:
+            hand_values = [hand.visible_points() for hand in next_hands]
+            starting_player = hand_values.index(max(hand_values))
+        next_valid_actions = next_hands[
```
Comment (noahbkim) [inline skyjo_immutable.py:299]: maybe needless to always compute next valid actions (constructing objects is expensive (though probably not relative to training))

**Review by noahbkim — APPROVED** [review]

File: skyjo_immutable.py [hunk 5]
```diff
@@ -70,6 +70,55 @@ def visible_indices(self) -> list[tuple[int, int]]:
...
+                and is_flipped[Hand.flat_index(r, col_idx)]
+                and cards[Hand.flat_index(0, col_idx)]
+                == cards[Hand.flat_index(r, col_idx)]
+            )
+        return all_same
+
+    def has_all_revealed(self) -> bool:
+        # To be cleared the is_flipped value would also be set to 1, so we can just
+        # count how many flipped cards there are without worrying about cleared counts.
+        return self.is_flipped.sum() == NUM_COLUMNS * NUM_ROWS
```
Comment (noahbkim) [inline skyjo_immutable.py:115]: I think u could probably just do `all(self.is_flipped)` and get short-circuiting

**Comment by urolyi** [comment]
Added Monte Carlo Tree Search, Model (SkyNet), and training code. There is a lot here and probably should've broken this up, but oh well.

I have a much better grasp of the training loop and overall algorithm structure, so happy to talk through it if needed. This is just the starting point and I'm sure all of the code will need evolve and improved, but it is running and vaguely seems to be improving the loss function when running.

Still need to add some testing for the MCTS and training code and also address some of the good suggestions you initially made for the actual immutable skyjo implementation. The main next milestones are figuring out the actual model architecture as currently it's just kind of a simple stand-in with only a single residual block. Then likely the most painful and time consuming part of running actual training loops and seeing if it works and debugging until it really seems to learn and become really good. There are some optimizations here that I'm curious to explore like in the training data generation, it is common to add multiple data points based on board symmetries, I'm wondering if we can do the same since each column is independent. Also, thinking about how we want to deal with the randomness present in the game. Maybe there is a way to add many new data points at each rng involved step instead of just a single outcome.

In case interested I heavily referenced a couple of repos for the training and mcts stuff so feel free to take a look as well.
- A very simple alpha-zero implementation for connect 2: https://github.com/JoshVarty/AlphaZeroSimple
- A general implementation for training alpha zero like models for games: https://github.com/suragnair/alpha-zero-general

File: test_skynet.py [hunk 6]
```diff
@@ -0,0 +1,228 @@
+import unittest
```
Comment (urolyi) [inline test_skynet.py:1]: This file was mostly generated with AI 😮 and it was actually pretty good imo
